### PR TITLE
Allow set operations for columns that map to same store type

### LINKED ISSUE #29020: SQL Server: Set operations should work if normalized store type is the same
In `SelectExpression.cs:1594` (Microsoft.EntityFrameworkCore.Query.SqlExpression) there is code that does
```
if (innerColumn1.TypeMapping!.StoreType != innerColumn2.TypeMapping!.StoreType)
{
    throw new InvalidOperationException(RelationalStrings.SetOperationsOnDifferentStoreTypes);
}
```

If `col1` type in mode is `VARCHAR(20)` and `col2` type is `varchar(20)` it throws. This should not happen as SQL server is not case sensitive. I udestrand why it would be desirable for `varchar(20)` and `varchar(50)` to throw, but in case of case sensitivity it's exactly the same type.

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #29020

### CODE REVIEW

File: src/EFCore/Storage/CoreTypeMapping.cs [hunk 1]
```diff
@@ -224,4 +224,14 @@ public virtual ValueComparer ProviderValueComparer
     /// <returns>An expression tree that can be used to generate code for the literal value.</returns>
     public virtual Expression GenerateCodeLiteral(object value)
         => throw new NotSupportedException(CoreStrings.LiteralGenerationNotSupported(ClrType.ShortDisplayName()));
+
+    /// <summary>
+    ///     Determines whether this type mapping is to the same store type as the given mapping.
+    /// </summary>
+    /// <param name="otherMapping">The other type mapping.</param>
+    /// <returns><see langword="true"/> if the two mappings map to the same type, <see langword="false"/> otherwise.</returns>
+    public virtual bool MapsToSameStoreType(CoreTypeMapping otherMapping)
```
Comment (roji) [inline src/EFCore/Storage/CoreTypeMapping.cs:233]: Should this only compare the actual (string) store type name into account? There could be different mappings - including even with different CLR types - which map to the same store type.

Comment (ajcvickers) [reply src/EFCore/Storage/CoreTypeMapping.cs:233]: Example?

Comment (ajcvickers) [reply src/EFCore/Storage/CoreTypeMapping.cs:233]: Never mind; I think you are right. Will re-work this. Was trying to be too clever. Always best not to try to be clever.

**Comment by stevendarby** [comment]
Why not just never throw that InvalidOperationException? If it proceeds and the query fails on the server because the store types aren't compatible then the user is in the same position: the query doesn't work. Currently this seems to be what is preventing set operations between, say, nbarchar(10) and nvarchar(20), which SQL Server is fine with.

**Comment by ajcvickers** [comment]
@roji @stevendarby Discussion is here: https://github.com/dotnet/efcore/pull/16813#discussion_r308851344

I'd be okay with removing this check entirely.

**Comment by roji** [comment]
Thanks for finding that discussion @ajcvickers. I'm also OK with removing the check.

**Comment by stevendarby** [comment]
@ajcvickers @roji Ah cool! I think that would close https://github.com/dotnet/efcore/issues/19129

**Comment by ajcvickers** [comment]
Decision from design meeting: stop throwing in this case and let the database throw if the types are not compatible.

**Comment by NetMage** [comment]
Is there a way to tell when this will be released (included in what version)?

**Comment by NetMage** [comment]
@ajcvickers Thank you, I was referring to the design meeting decision to stop throwing in this case.  I assume a new PR will have to be created to implement that decision, but the corresponding issues seem to have been closed based on this PR?

**Comment by ajcvickers** [comment]
https://github.com/dotnet/efcore/pull/30468

**Pull request closed without merging**